File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax/AccessorBlockSyntax.cs
using System;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
{
	internal sealed class AccessorBlockSyntax : MethodBlockBaseSyntax
	{
		internal readonly AccessorStatementSyntax _accessorStatement;

		internal readonly EndBlockStatementSyntax _endAccessorStatement;

		internal static Func<ObjectReader, object> CreateInstance;

		internal AccessorStatementSyntax AccessorStatement => _accessorStatement;

		internal EndBlockStatementSyntax EndAccessorStatement => _endAccessorStatement;

		public override MethodBaseSyntax Begin => AccessorStatement;

		public override EndBlockStatementSyntax End => EndAccessorStatement;

		internal AccessorBlockSyntax(SyntaxKind kind, AccessorStatementSyntax accessorStatement, GreenNode statements, EndBlockStatementSyntax endAccessorStatement)
			: base(kind, statements)
		{
			base._slotCount = 3;
			AdjustFlagsAndWidth(accessorStatement);
			_accessorStatement = accessorStatement;
			AdjustFlagsAndWidth(endAccessorStatement);
			_endAccessorStatement = endAccessorStatement;
		}

		internal AccessorBlockSyntax(SyntaxKind kind, AccessorStatementSyntax accessorStatement, GreenNode statements, EndBlockStatementSyntax endAccessorStatement, ISyntaxFactoryContext context)
			: base(kind, statements)
		{
			base._slotCount = 3;
			SetFactoryContext(context);
			AdjustFlagsAndWidth(accessorStatement);
			_accessorStatement = accessorStatement;
			AdjustFlagsAndWidth(endAccessorStatement);
			_endAccessorStatement = endAccessorStatement;
		}

		internal AccessorBlockSyntax(SyntaxKind kind, DiagnosticInfo[] errors, SyntaxAnnotation[] annotations, AccessorStatementSyntax accessorStatement, GreenNode statements, EndBlockStatementSyntax endAccessorStatement)
			: base(kind, errors, annotations, statements)
		{
			base._slotCount = 3;
			AdjustFlagsAndWidth(accessorStatement);
			_accessorStatement = accessorStatement;
			AdjustFlagsAndWidth(endAccessorStatement);
			_endAccessorStatement = endAccessorStatement;
		}

		internal AccessorBlockSyntax(ObjectReader reader)
			: base(reader)
		{
			base._slotCount = 3;
			AccessorStatementSyntax accessorStatementSyntax = (AccessorStatementSyntax)reader.ReadValue();
			if (accessorStatementSyntax != null)
			{
				AdjustFlagsAndWidth(accessorStatementSyntax);
				_accessorStatement = accessorStatementSyntax;
			}
			EndBlockStatementSyntax endBlockStatementSyntax = (EndBlockStatementSyntax)reader.ReadValue();
			if (endBlockStatementSyntax != null)
			{
				AdjustFlagsAndWidth(endBlockStatementSyntax);
				_endAccessorStatement = endBlockStatementSyntax;
			}
		}

		internal override void WriteTo(ObjectWriter writer)
		{
			base.WriteTo(writer);
			writer.WriteValue(_accessorStatement);
			writer.WriteValue(_endAccessorStatement);
		}

		static AccessorBlockSyntax()
		{
			CreateInstance = (ObjectReader o) => new AccessorBlockSyntax(o);
			ObjectBinder.RegisterTypeReader(typeof(AccessorBlockSyntax), (ObjectReader r) => new AccessorBlockSyntax(r));
		}

		internal override SyntaxNode CreateRed(SyntaxNode parent, int startLocation)
		{
			return new Microsoft.CodeAnalysis.VisualBasic.Syntax.AccessorBlockSyntax(this, parent, startLocation);
		}

		internal override GreenNode GetSlot(int i)
		{
			return i switch
			{
				0 => _accessorStatement, 
				1 => _statements, 
				2 => _endAccessorStatement, 
				_ => null, 
			};
		}

		internal override GreenNode SetDiagnostics(DiagnosticInfo[] newErrors)
		{
			return new AccessorBlockSyntax(base.Kind, newErrors, GetAnnotations(), _accessorStatement, _statements, _endAccessorStatement);
		}

		internal override GreenNode SetAnnotations(SyntaxAnnotation[] annotations)
		{
			return new AccessorBlockSyntax(base.Kind, GetDiagnostics(), annotations, _accessorStatement, _statements, _endAccessorStatement);
		}

		public override VisualBasicSyntaxNode Accept(VisualBasicSyntaxVisitor visitor)
		{
			return visitor.VisitAccessorBlock(this);
		}
	}
}

File: Microsoft.CodeAnalysis.VisualBasic/Microsoft.CodeAnalysis/VisualBasic/Syntax/InternalSyntax/AccessorBlockSyntax.vb
Imports Microsoft.CodeAnalysis
Imports Microsoft.CodeAnalysis.VisualBasic
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
Imports Roslyn.Utilities
Imports System
Imports System.Runtime.CompilerServices

Namespace Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
	Friend NotInheritable Class AccessorBlockSyntax
		Inherits Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.MethodBlockBaseSyntax
		Friend ReadOnly _accessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax

		Friend ReadOnly _endAccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax

		Friend Shared CreateInstance As Func(Of ObjectReader, Object)

		Friend ReadOnly Property AccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax
			Get
				Return Me._accessorStatement
			End Get
		End Property

		Public Overrides ReadOnly Property Begin As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.MethodBaseSyntax
			Get
				Return Me.AccessorStatement
			End Get
		End Property

		Public Overrides ReadOnly Property [End] As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax
			Get
				Return Me.EndAccessorStatement
			End Get
		End Property

		Friend ReadOnly Property EndAccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax
			Get
				Return Me._endAccessorStatement
			End Get
		End Property

		Shared Sub New()
			Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax.CreateInstance = Function(o As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax(o)
			ObjectBinder.RegisterTypeReader(GetType(Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax), Function(r As ObjectReader) New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax(r))
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal accessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax, ByVal statements As GreenNode, ByVal endAccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)
			MyBase.New(kind, statements)
			MyBase._slotCount = 3
			MyBase.AdjustFlagsAndWidth(accessorStatement)
			Me._accessorStatement = accessorStatement
			MyBase.AdjustFlagsAndWidth(endAccessorStatement)
			Me._endAccessorStatement = endAccessorStatement
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal accessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax, ByVal statements As GreenNode, ByVal endAccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax, ByVal context As ISyntaxFactoryContext)
			MyBase.New(kind, statements)
			MyBase._slotCount = 3
			MyBase.SetFactoryContext(context)
			MyBase.AdjustFlagsAndWidth(accessorStatement)
			Me._accessorStatement = accessorStatement
			MyBase.AdjustFlagsAndWidth(endAccessorStatement)
			Me._endAccessorStatement = endAccessorStatement
		End Sub

		Friend Sub New(ByVal kind As SyntaxKind, ByVal errors As DiagnosticInfo(), ByVal annotations As SyntaxAnnotation(), ByVal accessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax, ByVal statements As GreenNode, ByVal endAccessorStatement As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)
			MyBase.New(kind, errors, annotations, statements)
			MyBase._slotCount = 3
			MyBase.AdjustFlagsAndWidth(accessorStatement)
			Me._accessorStatement = accessorStatement
			MyBase.AdjustFlagsAndWidth(endAccessorStatement)
			Me._endAccessorStatement = endAccessorStatement
		End Sub

		Friend Sub New(ByVal reader As ObjectReader)
			MyBase.New(reader)
			MyBase._slotCount = 3
			Dim accessorStatementSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorStatementSyntax)
			If (accessorStatementSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(accessorStatementSyntax)
				Me._accessorStatement = accessorStatementSyntax
			End If
			Dim endBlockStatementSyntax As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax = DirectCast(reader.ReadValue(), Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.EndBlockStatementSyntax)
			If (endBlockStatementSyntax IsNot Nothing) Then
				MyBase.AdjustFlagsAndWidth(endBlockStatementSyntax)
				Me._endAccessorStatement = endBlockStatementSyntax
			End If
		End Sub

		Public Overrides Function Accept(ByVal visitor As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxVisitor) As Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.VisualBasicSyntaxNode
			Return visitor.VisitAccessorBlock(Me)
		End Function

		Friend Overrides Function CreateRed(ByVal parent As SyntaxNode, ByVal startLocation As Integer) As SyntaxNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.AccessorBlockSyntax(Me, parent, startLocation)
		End Function

		Friend Overrides Function GetSlot(ByVal i As Integer) As Microsoft.CodeAnalysis.GreenNode
			Dim greenNode As Microsoft.CodeAnalysis.GreenNode
			Select Case i
				Case 0
					greenNode = Me._accessorStatement
					Exit Select
				Case 1
					greenNode = Me._statements
					Exit Select
				Case 2
					greenNode = Me._endAccessorStatement
					Exit Select
				Case Else
					greenNode = Nothing
					Exit Select
			End Select
			Return greenNode
		End Function

		Friend Overrides Function SetAnnotations(ByVal annotations As SyntaxAnnotation()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax(MyBase.Kind, MyBase.GetDiagnostics(), annotations, Me._accessorStatement, Me._statements, Me._endAccessorStatement)
		End Function

		Friend Overrides Function SetDiagnostics(ByVal newErrors As DiagnosticInfo()) As GreenNode
			Return New Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax.AccessorBlockSyntax(MyBase.Kind, newErrors, MyBase.GetAnnotations(), Me._accessorStatement, Me._statements, Me._endAccessorStatement)
		End Function

		Friend Overrides Sub WriteTo(ByVal writer As ObjectWriter)
			MyBase.WriteTo(writer)
			writer.WriteValue(DirectCast(Me._accessorStatement, IObjectWritable))
			writer.WriteValue(DirectCast(Me._endAccessorStatement, IObjectWritable))
		End Sub
	End Class
End Namespace